Solver: shorten the skipping message if needed

erikd opened the pull request:
Also add a test to prevent this regressing.

Closes: https://github.com/haskell/cabal/issues/4251

Please read [Github PR Conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#github-pull-request-conventions) and then fill in *one* of these two templates.

---

**Template Α: This PR modifies [behaviour or interface](https://github.com/cabalism/cabal/blob/master/CONTRIBUTING.md#changelog)**

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
  * [ ] [Is the change significant?](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#is-my-change-significant) If so, remember to add `significance: significant` in the changelog file.
* [ ] The documentation has been updated, if necessary.
* [ ] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
* [x] Tests have been added. (*Ask for help if you don’t know how to write them! Ask for an exemption if tests are too complex for too little coverage!*)

erikd commented on cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:
```diff
@@ -282,16 +282,16 @@ showOption qpn@(Q _pp pn) (POption i linkedTo) =
-  | x@(POption i linkedTo) <- xs
-  ])
+  | x@(POption i linkedTo) <- take 3 xs
+  ] ++ if length xs >= 3 then " and earlier versions" else "")
```
If I use a number less that `3` here, a number of the `UnitTest` test will fail.

ulysses4ever replied:
So, when length of xs is exactly 3, cabal will print the entirety of the list but still says "and earlier versions"?..

3 looks like too many still. I'd suggest print 1 and fix the unit tests.

erikd committed:
Solver: shorten the skipping message if needed

Also add a test to prevent this regressing.

Closes: https://github.com/haskell/cabal/issues/4251

sebright reviewed (commented):
I have a few comments:

1. I think that -v3 should continue to show all of the versions, for debugging or when users need to see which versions cabal considered or their order. Then -v3 could also be used by any existing unit tests that would be broken by this change.
1. It looks like this change would shorten the skipping/rejecting message even when the solver doesn't reject all of the versions. Is that intentional, or did you mean to implement something more like #10926? Another idea is to shorten the list when the solver rejects/skips all of the remaining versions (from the current version to the least preferred version).
1. It might be better to continue listing any installed or linked versions, no matter where they appear in the list.
1. I agree with @ulysses4ever that only one version should be listed, at least when they are all non-linked source versions.
1. There should also be a test for shortening the "rejecting" message.
1. It would help to have a test where the solver rejects or skips multiple versions for one reason and then rejects another version for another reason. This test case would address (2).

erikd committed:
Change earlier to other in skipped versions msg

erikd committed:
Plumb verbosity level down to where its needed

Use `setVerbose` in two of the unit tests in preparation for upcoming
changes to the `skipping ...` message.

erikd commented:
>     1. I think that -v3 should continue to show all of the versions, for debugging or when users need to see which versions cabal considered or their order. Then -v3 could also be used by any existing unit tests that would be broken by this change.

I have this in my working tree.

>     2. It looks like this change would shorten the skipping/rejecting message even when the solver doesn't reject all of the versions. Is that intentional, or did you mean to implement something more like #10926? Another idea is to shorten the list when the solver rejects/skips all of the remaining versions (from the current version to the least preferred version).

Was not even aware that my change would also affect successful solutions. IMO, it should only affect the output when there is no successful solution. I will need some pointers to move forward on this.

>     3. It might be better to continue listing any installed or linked versions, no matter where they appear in the list.

Sorry, I do not even understand this comment.

>     4. I agree with @ulysses4ever that only one version should be listed, at least when they are all non-linked source versions.

Fixed in my local version already. Already have the full list being printed when `Verbosity` is set to `verbose` and had to add `setVerbose` to two of the unit tests, to disable the truncation of the list of skipped versions.

>     5. There should also be a test for shortening the "rejecting" message.

Any idea what the unit test for this would look like? I tried to add a test to `cabal-install/tests/UnitTests/Distribution/Solver/Modular/Solver.hs` but I cannot trigger it.

>     6. It would help to have a test where the solver rejects or skips multiple versions for one reason and then rejects another version for another reason. This test case would address (2).

Clues?

erikd commented:
Obviously these commtis should all be rebased down to a single commit. I am keeping this separate for now as each commit addresses a separate review comment.

geekosaur commented:
That's what `squash+merge me` is for.

erikd commented:
I know that `squahs+merge` can do it, but in the past I have seen it give poor results in comparison to doing it manually.

sebright reviewed (commented):
I realized that this PR is already a standalone improvement to the rejecting message, so I think it would be better to merge it before implementing the other parts of #10926 that I mentioned in my previous comment.

> > There should also be a test for shortening the "rejecting" message.
> 
> Any idea what the unit test for this would look like? I tried to add a test to `cabal-install/tests/UnitTests/Distribution/Solver/Modular/Solver.hs` but I cannot trigger it.

You could use a global constraint (equivalent to the `--constraint` flag) instead of a `build-depends` constraint, since global constraints haven't been optimized, and each version needs to be rejected separately.

I haven't reviewed the new changes yet, and I'm busy this week, so I probably won't have time for another review until later next week.

erikd committed:
Reduce show skipped versions from 3 to 1

And fix associated unit test expected string.

sebright reviewed (commented):
I read the rest of the code, and I only have a few small comments to add to my last review.

erikd committed:
Update 'and other versions' message

erikd committed:
Update changelog message

erikd commented on cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:
```diff
@@ -282,16 +283,16 @@ showOption qpn@(Q _pp pn) (POption i linkedTo) =
 -- >>> showOptions foobarQPN [k1, k2]
 -- "foo-bar; foo-bar~>bazqux.foo-bar-1, foo-bar~>bazqux.foo-bar-2"
 -- >>> showOptions foobarQPN [v0, i1, k2]
--- "foo-bar; 0, 1/installed-inplace, foo-bar~>bazqux.foo-bar-2"
```
Sorry, how do I manipulate the verbosity level in a doctest?

sebright replied:
You could probably use something like `mkVerbosity Verbose` to provide the verbosity argument.

erikd commented on cabal-install/tests/UnitTests/Distribution/Solver/Modular/Solver.hs:
```diff
@@ -962,8 +963,21 @@ tests =
+                    , Right $ exAv "B" 1 [ExFix "A" 4]
+                    ]
+                  rejecting = "rejecting: A-3.0.0 (conflict: B => A==4.0.0)"
+                  skipping = "skipping: A; 2.0.0 and other versions (has"
```
Add the number of other versions to the message.

erikd committed:
verbosity
